usermods/ULC_Battery_Management/usermod_ulc_batterymanagement.h

```diff
@@ -0,0 +1,279 @@
+#include "IP5306_I2C.h"
+#include <driver/adc.h>
+
+// the frequency to check battery voltage, 5 seconds
+#ifndef USERMOD_BATTERY_MEASUREMENT_INTERVAL
+  #define USERMOD_BATTERY_MEASUREMENT_INTERVAL 5000
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_PRECISION
+  #define USERMOD_BATTERY_ADC_PRECISION 4095.0f
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_REF
+  #define USERMOD_BATTERY_ADC_REF 3.3f
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_R1
+  #define USERMOD_BATTERY_ADC_R1 8.0f
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_R2
+  #define USERMOD_BATTERY_ADC_R2 2.5f
+#endif
+
+#ifndef USERMOD_BATTERY_MIN_VOLTAGE
+  #define USERMOD_BATTERY_MIN_VOLTAGE 2.6f
+#endif
+
+#ifndef USERMOD_BATTERY_MAX_VOLTAGE
+  #define USERMOD_BATTERY_MAX_VOLTAGE 4.2f
+#endif
+
+#ifndef USERMOD_IP5306_SDA
+  #define USERMOD_IP5306_SDA 32
+#endif
+
+#ifndef USERMOD_IP5306_SCL
+  #define USERMOD_IP5306_SCL 33
+#endif
+
+#ifndef USERMOD_ABL_BATTERY
+  #define USERMOD_ABL_BATTERY 1250
+#endif
+
+#ifndef USERMOD_ABL_EXT
+  #define USERMOD_ABL_EXT 2500
+#endif
+
+class UsermodULCBatteryManagement : public Usermod {
+
+  private:
+    IP5306 *ip5306 = nullptr;
+
+    bool initDone = false;
+    bool initializing = true;
+    // GPIO pin used for battery measurment (with a default compile-time fallback)
+    int8_t ip5306_sda = USERMOD_IP5306_SDA;
+    int8_t ip5306_scl = USERMOD_IP5306_SCL;
+    // how often do we measure?
+    unsigned long readingInterval = USERMOD_BATTERY_MEASUREMENT_INTERVAL;
+    unsigned long nextReadTime = 0;
+    unsigned long lastReadTime = 0;
+
+    uint32_t ablBattery = USERMOD_ABL_BATTERY;
+    uint32_t ablExt = USERMOD_ABL_EXT;
+    
+    // battery min. voltage
+    float minBatteryVoltage = USERMOD_BATTERY_MIN_VOLTAGE;
+    // battery max. voltage
+    float maxBatteryVoltage = USERMOD_BATTERY_MAX_VOLTAGE;
+
+    uint8_t batteryLevel = 0; 
+    bool isCharging = false;
+    bool isPluggedIn = false;
+
+    // strings to reduce flash memory usage (used more than twice)
+    static const char _name[];
+    static const char _readInterval[];
+    static const char _maxCurrentBattery[];
+    static const char _maxCurrentExt[];
+
+    // custom map function
+    // https://forum.arduino.cc/t/floating-point-using-map-function/348113/2
+    double mapf(double x, double in_min, double in_max, double out_min, double out_max) 
+    {
+      return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+    }
+
+    float truncate(float val, byte dec) 
+    {
+      float x = val * pow(10, dec);
+      float y = round(x);
+      float z = x - y;
+      if ((int)z == 5)
+      {
+          y++;
+      }
+      x = y / pow(10, dec);
+      return x;
+    }
+
+  public:
+
+    void setup() {
+      DEBUG_PRINTLN(F("Allocating ip5306 pins..."));
+      PinManagerPinType pins[2] = { { ip5306_sda, true }, { ip5306_scl, true } };
+      if (pinManager.allocateMultiplePins(pins, 2, PinOwner::HW_I2C))
+      {
+        ip5306 = new IP5306(ip5306_sda, ip5306_scl);
```

Comment:
For I2C bus sharing, its very important that all usermods rely on the global `Wire` object.

Please check if there is a variant of the IP5306 constructor that takes a reference to `Wire`, instead of getting SDA and SCL pins as parameters.

Reply:
You can take a look at the BME280 usermod as an example.
https://github.com/Aircoookie/WLED/blob/c04c73bbd7f448312c368e33773e510d4e8324cf/usermods/BME280_v2/usermod_bme280.h#L187-L195

- remove I2C pins from the usermod settings
- use global i2c_sda and i2c_sda if you need them in your code - read only!
- Leave out Wire.begin() as `Wire` is already initialized for you by WLED
- check for `i2c_sda < 0 || i2c_scl < 0` and disable the usermod in case the global pins are undefined (-1)


@blazoncek do you know a simpler example that shows the new concept?

Reply:
I have switched to the new i2c object now. However i have the issue that i cannot set default i2c pins for the board in the platformio.ini. If i use HW_PIN_SDA it does not work at all (and the pin in usermods is "unused") if i use I2CSDAPIN it shows in the usermods but only works after saving the settings again.

All testing was done with a factory reset of wled

Reply:
Checked the code and it looks like not all instances of I2C bus were correctly updated. My mistake.
Actually there is no need to allocate any pins as they are allocated by WLED core at boot.

A usermod should only check if `i2c_sda` and `i2c_scl` are not equal to -1. In such case I2C interface is initialised and `Wire` global started.

Reply:
I think we are talking about seperate things. I need to set the i2c pins for my board in the target configuration. But it does not use the values I give in the target configuration

Reply:
> I have switched to the new i2c object now. However i have the issue that i cannot set default i2c pins for the board in the platformio.ini. If i use HW_PIN_SDA it does not work at all (and the pin in usermods is "unused") if i use I2CSDAPIN it shows in the usermods but only works after saving the settings again.

If you want to allocate and assign (as opposed to only define) I2C pins on fresh install of WLED, then use `-D I2CSDAPIN=X` and `-D I2CSCLPIN=Y` if you just need to define them then use `-D HW_PIN_SDA=X` and `-D HW_PIN_SCL=Y`.
Defining pins is only informational to the user (placeholder text in UI). If you use I2CSDAPIN and I2CSCLPIN then those will overwrite HW_PIN_SDA & HW_PIN_SCL.

This is the relevant code executed at boot which will load compiled in defaults (I2CxxxPIN).
![Screenshot 2023-06-20 at 22 29 30](https://github.com/Aircoookie/WLED/assets/7722392/57702995-a670-4f71-9fde-ed66a6275460)

Reply:
Okay, yes using I2CSDA/SCLPIN works. But I still need to go to the Settings page and save once for it to actually work

Reply:
I have hard time believing that. 😀

Reply:
Well then, how do I make you believe, or do you have any other Idea why it does this?